### PR TITLE
fix: Add supported dict typing for client_options

### LINKED ISSUE #1380: Dictionary support not reflected in type annotations
## Issue

- This is a follow-up to [#986](https://github.com/googleapis/gapic-generator-python/issues/986) (closed). Retesting the recent `google-cloud-documentai==1.5.0` version, the issue is still present.
- This code is valid:
  ```
  import google.cloud.documentai_v1 as documentai

  client_options = dict(api_endpoint="eu-documentai.googleapis.com")
  client = documentai.DocumentProcessorServiceClient(client_options=client_options)
  ```
- But it triggers the following type checker warning:
  ```
  Argument of type "dict[str, str]" cannot be assigned to parameter "client_options" of type "ClientOptions | None" in function "__init__"
  Type "dict[str, str]" cannot be assigned to type "ClientOptions | None"
  "dict[str, str]" is incompatible with "ClientOptions"
  ```
- If you check the [source](https://github.com/googleapis/python-documentai/blob/1b4eff4e0f6ca34bd12fb5cd01343ca74b4f9633/google/cloud/documentai_v1/services/document_processor_service/client.py#L366), `client_options` can be a dict, yet its declaration says otherwise:
  ```python
  def __init__(
      ...
      client_options: Optional[client_options_lib.ClientOptions] = None,
      ...
  ) -> None:
      ...
      if isinstance(client_options, dict):
          client_options = client_options_lib.from_dict(client_options)
  ```

## Expected

  ```python
  def __init__(
      ...
      client_options: Union[client_options_lib.ClientOptions, dict, None] = None,
      ...
  ) -> None:
      ...
  ```

### PR DESCRIPTION
**Pull request by PicardParis**
Fixes #1380

### CODE REVIEW

**Comment by PicardParis** [comment]
Hi @parthea,
Remade this simple PR first.
Can you launch the tests & review?

File: gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 1]
```diff
@@ -325,6 +325,7 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
             client_options = client_options_lib.from_dict(client_options)
         if client_options is None:
             client_options = client_options_lib.ClientOptions()
+        client_options = cast(client_options_lib.ClientOptions, client_options)
```
[the remark this answers is not in the document]
Comment (PicardParis) [reply gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:328]: done

**Comment by PicardParis** [comment]
@parthea, can you relaunch the tests? (not available on my side)

**Review by parthea — APPROVED** [review]

**Comment by PicardParis** [comment]
- Sync'ed with the main branch.
- Auto-tests are ok.
- `OwlBot Post Processor` to be launched manually.

**Review by parthea — APPROVED** [review]

**Pull request merged**